spatialos-sdk/src/worker/component.rs

```diff
@@ -1,35 +1,463 @@
+        pub schema_type: SchemaComponentUpdate,
+        pub user_handle: *mut Worker_ComponentUpdateHandle,
+    }
+
+    impl From<&Worker_ComponentUpdate> for ComponentUpdate {
+        fn from(update: &Worker_ComponentUpdate) -> Self {
+            ComponentUpdate {
+                component_id: update.component_id,
+                schema_type: SchemaComponentUpdate {
+                    component_id: update.component_id,
+                    internal: update.schema_type,
+                },
+                user_handle: update.user_handle,
+            }
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct CommandRequest {
+        pub component_id: ComponentId,
+        pub schema_type: SchemaCommandRequest,
+        pub user_handle: *mut Worker_CommandRequestHandle,
+    }
+
+    impl From<&Worker_CommandRequest> for CommandRequest {
+        fn from(request: &Worker_CommandRequest) -> Self {
+            CommandRequest {
+                component_id: request.component_id,
+                schema_type: SchemaCommandRequest {
+                    component_id: request.component_id,
+                    internal: request.schema_type,
+                },
+                user_handle: request.user_handle,
+            }
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct CommandResponse {
+        pub component_id: ComponentId,
+        pub schema_type: SchemaCommandResponse,
+        pub user_handle: *mut Worker_CommandResponseHandle,
+    }
+
+    impl From<&Worker_CommandResponse> for CommandResponse {
+        fn from(response: &Worker_CommandResponse) -> Self {
+            CommandResponse {
+                component_id: response.component_id,
+                schema_type: SchemaCommandResponse {
+                    component_id: response.component_id,
+                    internal: response.schema_type,
+                },
+                user_handle: response.user_handle,
+            }
         }
     }
 }
 
-// TODO: Wrap Schema_ComponentUpdate
-pub struct ComponentUpdate {
-    pub component_id: ComponentId,
-    pub schema_type: *mut Schema_ComponentUpdate,
+// A data structure which represents all known component types. Used to generate an array of vtables to pass
+// to the connection object.
+pub struct ComponentDatabase {
+    component_vtables: Vec<worker::Worker_ComponentVtable>,
+}
+
+impl ComponentDatabase {
+    pub fn new() -> Self {
+        ComponentDatabase {
+            component_vtables: Vec::new(),
+        }
+    }
+
+    pub fn add_component<C: Component>(mut self) -> Self {
+        self.component_vtables.push(create_component_vtable::<C>());
+        self
+    }
+
+    pub(crate) fn to_worker_sdk(&self) -> *const worker::Worker_ComponentVtable {
+        self.component_vtables.as_ptr()
+    }
+
+    pub(crate) fn len(&self) -> usize {
+        self.component_vtables.len()
+    }
+}
+
+// An object which contains a reference counted instance to T.
+struct ClientHandle<T> {
+    data: Rc<T>,
```

Comment:
`data` should probably an `Arc<T>` since the data will be sent to the networking thread and freed from there (if I'm understanding the serialization docs correctly).